aten/src/ATen/native/LinearAlgebra.cpp

```diff
@@ -1664,101 +1666,119 @@ Tensor _matmul_impl(
+    const auto t1 = transpose ? MaybeOwned<Tensor>::owned(tensor2.mT())
+                              : MaybeOwned<Tensor>::borrowed(tensor1);
+    const auto t2 = !transpose ? MaybeOwned<Tensor>::borrowed(tensor2)
+                               : dim_tensor1 == 2
+                                   ? MaybeOwned<Tensor>::owned(tensor1.t())
+                                   : MaybeOwned<Tensor>::borrowed(tensor1);
+    // Invariant: t1->dim() >= 3 && (t2->dim() == 1 || t2->dim() == 2)
+    //            and *t1 and *t2 are matmul-compatible
+
+    // Why not t1->view(-1, sizes_1.back())?
     // If the last dim is 0, then view(-1, 0) won't work because the -1 becomes ambiguous.
     // This can happen in e.g. [3, 5, 0] @ [0, 0].
-    // So we manually compute the folding as a result.
-    const auto dim1_size = c10::multiply_integers(size1.begin(), size1.end() - 1);
-    auto t1 = tensor1.expect_contiguous()->view({dim1_size, size1[size1.size() - 1]});
-    Tensor output = has_out ? at::_unsafe_view(at::mm_out(out, t1, t2), output_size)
-                            : at::_unsafe_view(t1.mm(t2), output_size);
-    return has_out ? out.set_(output) : output;
-  } else if ((dim_tensor1 == 1 || dim_tensor1 == 2) && dim_tensor2 >= 3) {
-    // optimization: transpose the inner dimensions of the arguments, call
-    // matmul on the swapped arguments, then transpose the inner dimensions
-    // of the result.
-    const int64_t n = dim_tensor1 == 2 ? tensor1.size(-2) : 1;
-    const int64_t m = tensor1.size(-1);
-    const int64_t p = tensor2.size(-1);
-
-    const Tensor t2_T = tensor2.transpose(-1, -2);
-    const Tensor t1_T = dim_tensor1 == 2 ? tensor1.t() : tensor1.reshape({n, m}).t();
-    const Tensor res_T = _matmul_impl(out, t2_T, t1_T);
-
-    if (dim_tensor1 == 2) {
-      Tensor res = res_T.transpose(-1, -2).contiguous();
-      return has_out ? out.set_(res) : res;
+    const auto sizes_1 = t1->sizes();
+    auto output_shape = DimVector(sizes_1.begin(), sizes_1.end() - 1);
+    const auto folded_dim1 = c10::multiply_integers(output_shape);
+
+    // Readjust output_shape if we are multiplying by a matrix
+    const auto t2_is_matrix = t2->dim() == 2;
+    if (t2_is_matrix) {
+      output_shape.push_back(t2->sizes()[1]);
     }
-    else {
-      c10::IntArrayRef shape_array = tensor2.sizes().slice(0, dim_tensor2 - 2);
-      DimVector shape(shape_array.begin(), shape_array.end());
-      shape.push_back(p);
-
-      Tensor res = res_T.reshape(shape).contiguous();
-      return has_out ? out.set_(res) : res;
+    const auto t1_folded = t1->reshape({folded_dim1, sizes_1.back()});
+    if (!has_out) {
+      if (t2_is_matrix) {
+        // FIXME This path always does an unnecessary copy when transpose == true as the returned
+        // result from BLAS is already C-transposed
+        const auto output = at::_unsafe_view(t1_folded.mm(*t2), output_shape);
+        return transpose ? output.transpose_(-2, -1).contiguous() : output;
```

Comment:
again, as soon as you incur additional copy 3d-2d becomes not much of an optimization, bmm is likely better.

Reply:
We could still go through this path if we called `mm_out` with a matrix transposed strides, but that'd make `matmul` a non-composite.
Should we just dispatch to `bmm` in this case?

Reply:
Yeah dispatching to bmm sounds good, and same should happen for `_out` cases, having different paths for `out` and non-out is super confusing.